addons/mass_mailing/models/mass_mailing.py

```diff
@@ -54,6 +54,50 @@ def _compute_contact_nbr(self):
         for mailing_list in self:
             mailing_list.contact_nbr = data.get(mailing_list.id, 0)
 
+    @api.multi
+    def copy(self, default=None):
+        recipients = self.env['mail.mass_mailing.contact'].search([('list_ids', 'in', self.ids), ('opt_out', '!=', True)])
+        mailing_list = super(MassMailingList, self).copy(default=default)
+        for recipient in recipients:
+            recipient.write({'list_ids': [(4, mailing_list.ids)]})
+        return mailing_list
+
+    def merge_massmail_list(self, mailing_lists):
+        self.env.cr.execute("""
```

Comment:
While you are at it do a single sql query. It seems strange to optimize the search and not optimize the write.